### DISCUSSION TITLE
Event Queue Documentation

### DISCUSSION THREAD

**Discussion by anthonykasza** [zeek/zeek#1162] [Archive]
I am searching for documentation on Zeek's event queue. Specifically, I'm interested in the different states Zeek transitions through from invocation to termination. My current mental model is:

1. Scripts are parsed, starting at some entry point. I believe zeek-init is typically this entry point.
2. @load directives are evaluated and #1 is visited again until all scripts are loaded
3. Remaining directives are evaluated
4. `zeek_init` is handled
5. ?
6. `zeek_done` is handled
7. The zeek process terminates

Currently, the way I determine what the event engine will do is to write a script and seeing what happens. The "Bro Architecture" diagram from Robin's 2017 presentation, "Following the Packets: A Walk Through Bro’s Internal Processing Pipeline" is a very helpful overview. I'm looking for more depth around the event queue component. 

Thank you for your time and response.

**Comment by 0xxon** [comment]
I am not quite sure if this is what you are looking for, but...

typically the event engine just raises events in the order that they are put into it. So - after Zeek is started up, it will start handling the iosources. That typically means that network traffic is sent to the analyzers. Events are queued in the order that the analyzers (or - scripts) raise them - and then executed in the order that they were queued.

The order that analyzers put items in is up to the individual analyzers - so there the best way to figure out ordering is probably really to try. Generally speaking things will be raised in the order that they are seen in traffic - but if several events are raised for one packet, this becomes a bit unclear (and you just have to know what the analyzer does internally).

In addition there are timers that perform actions - like, e.g., scheduled events, or removal of expired items from lists, etc.. These are called once they are due (because the time advanced to the point where they fire).

Does that answer your question?

Generally - a lot of internals are not documented very well - and if you need that level of knowledge you will currently either need to ask around, or dig through the source.

Since this is not a bug, I am going to close this. Questions like these are generally better for the mailing list or Slack. Feel free to re-open this if you want this to be a documentation issue (so - if the complaint is "there is no documentation about this and there should be").

**Comment by vpax** [comment]
I think @anthonykasza meant this as an issue, not as a request that someone give him an explanation.  I.e., this should be documented clearly, and as best as he could tell (from our conversation about this earlier) he couldn't find it.

**Comment by vpax** [comment]
Also, you sketched events in general.  He means the special events associated with execution states, like zeek_init, zeek_done, scripts_parsed (something like that)

**Comment by anthonykasza** [comment]
Correct, @vpax.
My apologies for filing this incorrectly, @0xxon. I was hoping for either a link to existing documentation or for documentation to be created.

**Comment by rsmmr** [comment]
I'm going to close this as it's not a specific issue and just a piece of broader lack of docs for internals. If we want to discuss further how to improve, I suggest moving over to Discussions.
